workspaces in pipeline start modal

abhinandan13jan opened the pull request:
# Addresses
https://issues.redhat.com/browse/ODC-4772

# Screenshot
![Screencast from 03-31-2021 07_59_39 PM](https://user-images.githubusercontent.com/24852534/113164869-21b52480-925f-11eb-8f16-d2d0e4d5ac4f.gif)

# Tests
Added VolumeClaimTemplate.spec

# Browser Conformance
Chrome,

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,179 @@
+          let radioObj = null;
+          const disabled = !allowedAccessModes.includes(radio.value);
+
+          allowedAccessModes.forEach((mode) => {
+            const checked = !disabled ? radio.value === accessMode : radio.value === mode;
+            radioObj = (
+              <RadioInput
+                {...radio}
+                key={radio.value}
+                onChange={handleAccessMode}
+                inline
+                disabled={disabled}
+                checked={checked}
+                aria-describedby="access-mode-help"
+                name="accessMode"
+              />
+            );
```
Doing a `.forEach` in the middle of rendering is mighty weird... are you not just going to override your rendering multiple times before rendering?

abhinandan13jan replied:
Moved the forEach out and only applying to set the `checked` flag. I am replicating the same logic used in createPVCForm.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,179 @@
+    setAccessMode(event.currentTarget.value);
+  };
+
+  const handleStorageClass = (updatedStorageClass) => {
+    const provisioner: string = updatedStorageClass?.provisioner || '';
+    // if the provisioner is unknown or no storage class selected, user should be able to set any access mode
+    const modes = provisionerAccessModeMapping[provisioner]
+      ? provisionerAccessModeMapping[provisioner]
+      : getAccessModeForProvisioner(provisioner);
+    // setting message to display for various modes when a storage class of a know provisioner is selected
+    const displayMessage =
+      provisionerAccessModeMapping[provisioner] || isCephProvisioner(provisioner)
+        ? 'Access mode is set by storage class and cannot be changed'
+        : 'Permissions to the mounted drive';
+    setAccessMode('ReadWriteOnce');
+    setAccessModeHelp(displayMessage);
+    // setting accessMode to default with the change to Storage Class selection
+    setAllowedAccessModes(modes);
+    setStorageClass(updatedStorageClass?.metadata?.name);
+    setStorageProvisioner(provisioner);
+    if (storageProvisioner.includes(cephStorageProvisioners[1])) {
+      setVolumeMode('Filesystem');
+    }
+  };
```
This feels weird... I wonder if we need some interim handler or wrapper of the StorageClass to simplify this logic. It's a dropdown after all.

abhinandan13jan replied:
The same login applies for the CreatePVC dropdown. Hence, kept it. I dont think we can move it out as a util as it is directly changing the states of this component.

However, should we do this validation on client or should we let the server take care of this?

andrewballantyne replied:
I'll have to relook at this once you clean things up (other comments). It's possible there is a way to streamline this so it's not so heavy.

andrewballantyne replied:
Still need to check on this, but it should be fine as-is if I cannot figure out a better solution.

andrewballantyne reviewed (commented):
Not doing a full review or anything, but just reading the code gets me these comments. If they are part of the path you're working on, ignore away 🙂

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/PiplelineWorkspacesSection.tsx:
```diff
@@ -36,6 +38,20 @@ const getVolumeTypeFields = (volumeType: VolumeTypes, index: number, t: TFunctio
+        <VolumeClaimTemplateForm
+          name={`workspaces.${index}.data.volumeClaimTemplate`}
+          initialSizeValue="200"
+          initialSizeUnit="Mi"
```
yes, it is 1Gi, updated accordingly

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,179 @@
+  ]);
+
+  return (
+    <ExpandCollapse
```
So, there is a help-block above this now. added as per UX screenshot

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/PiplelineWorkspacesSection.tsx:
```diff
@@ -36,6 +38,20 @@ const getVolumeTypeFields = (volumeType: VolumeTypes, index: number, t: TFunctio
     }
+    case VolumeTypes.EmptyDirectory: {
+      return (
+        <Alert variant="info" title="this option could not be used to share data between tasks" />
```
```suggestion
        <Alert inline variant="info" title="this option could not be used to share data between tasks" />
```
We don't use non-inline versions anywhere in the console I don't think. At least not in the DevConsole (that I have seen)

We'll also need to translate this... but I suppose we can wait on UX wording.

andrewballantyne replied:
And the wording has been figured out [here on the designs](https://docs.google.com/document/d/12ZETv528Puq41o9zg8_GCvUhhd5g8NEiGbA3bnmawjs/edit?disco=AAAALzTmAK4).

abhinandan13jan replied:
updated

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,181 @@
+  return (
+    <>
+      <p className="help-block">
+        {t('pipelines-plugin~This will create a PersistentVolumeClaim with a size of 1 GiB')}
```
Maybe we use the actual value here... That way when they change it, the wording doesn't say the incorrect thing.

abhinandan13jan replied:
yes, that seems to be the better way. Updated accordingly

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/VolumeClaimTemplateForm.spec.tsx:
```diff
@@ -0,0 +1,69 @@
+
+describe('VolumeClaimTemplateForm', () => {
+  beforeEach(() => {
+    VolumeClaimTemplateFormProps = {
```
```suggestion
    volumeClaimTemplateFormProps = {
```
Even in tests we should still use camelCase for variable names.

abhinandan13jan replied:
yes

andrewballantyne commented:
![image](https://user-images.githubusercontent.com/8126518/111354594-db57a700-865c-11eb-8502-830e2e9f9f01.png)

Please align this with the design. Those labels are harsh standing out.

abhinandan13jan commented:
> ![image](https://user-images.githubusercontent.com/8126518/111354594-db57a700-865c-11eb-8502-830e2e9f9f01.png)
> 
> Please align this with the design. Those labels are harsh standing out.

@andrewballantyne Updated screenshot. Now using the same classes as that of create-pvc form

rottencandy commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,184 @@
+    // setting message to display for various modes when a storage class of a know provisioner is selected
+    const displayMessage =
+      provisionerAccessModeMapping[provisioner] || isCephProvisioner(provisioner)
+        ? 'Access mode is set by storage class and cannot be changed'
+        : 'Permissions to the mounted drive';
```
Should we translate these?

abhinandan13jan replied:
updated

karthikjeeyar commented:
Through git import flow, we are creating default volumeClaimTemplates behind the scenes, with a pipeline label attached to VolumentClaimTemplate, this label will be attached to the newly created PVC  as well so start modal can auto select an existing PVC for that pipeline. A user who starts the pipeline for the second time will see the existing PVC selected for them, I think we should also include this labeling of VolumeClaimTemplate here. @andrewballantyne WDYT?

https://github.com/openshift/console/blob/master/frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts#L133-L135

andrewballantyne commented:
> Through git import flow, we are creating default volumeClaimTemplates behind the scenes, with a pipeline label attached to VolumentClaimTemplate, this label will be attached to the newly created PVC as well so start modal can auto select an existing PVC for that pipeline. A user who starts the pipeline for the second time will see the existing PVC selected for them, I think we should also include this labeling of VolumeClaimTemplate here. @andrewballantyne WDYT?
> 
> https://github.com/openshift/console/blob/master/frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/utils.ts#L133-L135

I think that's a design question @karthikjeeyar - My thoughts are "sure, why not". @bgliwa01 Opinion?

nemesis09 commented:
Please run "yarn i18n" to fix failing tests

abhinandan13jan commented on frontend/public/locales/en/public.json:
```diff
@@ -637,5 +637,6 @@
   "Disabled": "Disabled",
-  "Enabled": "Enabled"
+  "Enabled": "Enabled",
+  "Empty Directory doesn't support shared data between tasks": "Empty Directory doesn't support shared data between tasks"
```
I missed the prefix. Updated now

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/PiplelineWorkspacesSection.tsx:
```diff
@@ -36,6 +38,24 @@ const getVolumeTypeFields = (volumeType: VolumeTypes, index: number, t: TFunctio
+        <Alert
+          isInline
+          variant="info"
+          title={t("Empty Directory doesn't support shared data between tasks")}
```
updated

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,184 @@
+      <p className="help-block">
+        {t(
+          'pipelines-plugin~This will create a PersistentVolumeClaim with a size of {{initialSizeValue}} {{initialSizeUnit}}',
+          { initialSizeValue, initialSizeUnit },
```
makes sense. updated accordingly

abhinandan13jan commented:
@andrewballantyne @karthikjeeyar @nemesis09 Thanks for the review. Updated accordingly

nemesis09 commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,184 @@
+            let checked = null;
+            allowedAccessModes.forEach((mode) => {
+              checked = !disabled ? radio.value === accessMode : radio.value === mode;
+            });
```
in the code piece 
` allowedAccessModes.forEach((mode) => {
     checked = !disabled ? radio.value === accessMode : radio.value === mode;
});`
wouldn't the expression `radio.value === mode` always yield false when disabled is true  i.e `allowedAccessModes` does not include radio.value?

nemesis09 replied:
could we instead write it as
`const checked = disabled ? disabled : radio === accessMode;`

abhinandan13jan replied:
What you say makes sense. But I've only replicated the logic used in create-pvc form without changing what they are trying to achieve.

I think in the case when all radios are disabled and one is selected by default. i.e user has no authority to make a change, this expression will hold true

abhinandan13jan replied:
updated. Works fine locally

nemesis09 replied:
I'm a bit confused how this will account for a default selection when every `radio` is disabled.
As far as i could understand, when `disabled` is true, every iteration of `forEach` evaluates the expression `radio.value === mode` which would never yield true since `disabled` is true only when `allowedAccessModes` does not include `radio.value`.

debsmita1 commented:
@abhinandan13jan you seem to have missed adding validation in the form, the `Size` field is made required but no error is shown when it is empty

![Screenshot from 2021-03-19 20-21-46](https://user-images.githubusercontent.com/22490998/111799131-be3b0680-88f0-11eb-9268-3c449af11a62.png)

nemesis09 commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/__tests__/VolumeClaimTemplateForm.spec.tsx:
```diff
@@ -0,0 +1,69 @@
+    };
+  });
+
+  it('should render correct HelpText and ExpandCollapse', () => {
```
should we be testing if a component renders unless it is rendered conditionally or the number of instances of a component rendered, changes?

abhinandan13jan commented:
> @abhinandan13jan you seem to have missed adding validation in the form, the `Size` field is made required but no error is shown when it is empty
> 
> ![Screenshot from 2021-03-19 20-21-46](https://user-images.githubusercontent.com/22490998/111799131-be3b0680-88f0-11eb-9268-3c449af11a62.png)

I am not sure, whether we should validate anything on the client end. The patched spec should be validated against the tekton API. cc: @andrewballantyne What should be done here?

nemesis09 commented:
> > @abhinandan13jan you seem to have missed adding validation in the form, the `Size` field is made required but no error is shown when it is empty
> > ![Screenshot from 2021-03-19 20-21-46](https://user-images.githubusercontent.com/22490998/111799131-be3b0680-88f0-11eb-9268-3c449af11a62.png)
> 
> I am not sure, whether we should validate anything on the client end. The patched spec should be validated against the tekton API. cc: @debsmita1 @andrewballantyne What should be done here?

Even if the spec is validated against the API, should we allow a submit with invalid values?
Currently the form submits with an empty "size" field and an error log is shown in the Log Snippet of the PipelineRun details page.
![Screenshot from 2021-03-19 22-46-00](https://user-images.githubusercontent.com/38663217/111818851-ab7efc80-8905-11eb-8992-7c59bcad1c4d.png)

abhinandan13jan commented:
> > > @abhinandan13jan you seem to have missed adding validation in the form, the `Size` field is made required but no error is shown when it is empty
> > > ![Screenshot from 2021-03-19 20-21-46](https://user-images.githubusercontent.com/22490998/111799131-be3b0680-88f0-11eb-9268-3c449af11a62.png)
> > 
> > 
> > I am not sure, whether we should validate anything on the client end. The patched spec should be validated against the tekton API. cc: @debsmita1 @andrewballantyne What should be done here?
> 
> Even if the spec is validated against the API, should we allow a submit with invalid values?
> Currently the form submits with an empty "size" field and an error log is shown in the Log Snippet of the PipelineRun details page.
> ![Screenshot from 2021-03-19 22-46-00](https://user-images.githubusercontent.com/38663217/111818851-ab7efc80-8905-11eb-8992-7c59bcad1c4d.png)

validation added

nemesis09 commented:
> > > > @abhinandan13jan you seem to have missed adding validation in the form, the `Size` field is made required but no error is shown when it is empty
> > > > ![Screenshot from 2021-03-19 20-21-46](https://user-images.githubusercontent.com/22490998/111799131-be3b0680-88f0-11eb-9268-3c449af11a62.png)
> > > 
> > > 
> > > I am not sure, whether we should validate anything on the client end. The patched spec should be validated against the tekton API. cc: @debsmita1 @andrewballantyne What should be done here?
> > 
> > 
> > Even if the spec is validated against the API, should we allow a submit with invalid values?
> > Currently the form submits with an empty "size" field and an error log is shown in the Log Snippet of the PipelineRun details page.
> > ![Screenshot from 2021-03-19 22-46-00](https://user-images.githubusercontent.com/38663217/111818851-ab7efc80-8905-11eb-8992-7c59bcad1c4d.png)
> 
> validation added

I tried it out but apparently it still allows submit with empty field and invalid values

nemesis09 commented:
I think the problem in the validation is that it is not validating "requestSize" alone and instead validating "storage" which is "requestSize"+"requestUnit"
so even with an empty "requestSize" field, storage does get a string containing only the "requestUnit", which is sufficient for a successful validation as per `yup.string().required()`

nemesis09 commented:
Thanks @abhinandan13jan 
The validations work fine now.
I just have some observations around the `help-block` message that is generated using the `requestSize` and `requestUnit`.
The message shows invalid or no values when the `requestSize` has such values.
here are some examples-

- `requestSize` is empty
![Screenshot from 2021-03-24 17-21-56](https://user-images.githubusercontent.com/38663217/112306421-cb2b7180-8cc5-11eb-9700-17aa5c8a56f6.png)
 
- `requestSize` has non numeric values
![Screenshot from 2021-03-24 17-22-09](https://user-images.githubusercontent.com/38663217/112306232-9ae3d300-8cc5-11eb-8ab0-ee5511614bf9.png)

- `requestSize` has negative numeric values
![Screenshot from 2021-03-24 17-22-25](https://user-images.githubusercontent.com/38663217/112306351-b949ce80-8cc5-11eb-8e61-7f9b85a75311.png)

cc @bgliwa01

abhinandan13jan commented:
/test frontend

abhinandan13jan commented:
> Thanks @abhinandan13jan
> The validations work fine now.
> I just have some observations around the `help-block` message that is generated using the `requestSize` and `requestUnit`.
> The message shows invalid or no values when the `requestSize` has such values.
> here are some examples-
> 
> * **`requestSize` is empty**
> 
> ![Screenshot from 2021-03-24 17-21-56](https://user-images.githubusercontent.com/38663217/112306421-cb2b7180-8cc5-11eb-9700-17aa5c8a56f6.png)
> 
> * **`requestSize` has non numeric values**
> 
> ![Screenshot from 2021-03-24 17-22-09](https://user-images.githubusercontent.com/38663217/112306232-9ae3d300-8cc5-11eb-8ab0-ee5511614bf9.png)
> 
> * **`requestSize` has negative numeric values**
> 
> ![Screenshot from 2021-03-24 17-22-25](https://user-images.githubusercontent.com/38663217/112306351-b949ce80-8cc5-11eb-8e61-7f9b85a75311.png)
> 
> cc @bgliwa01

updated please re-review

nemesis09 commented:
tested locally
works fine
values such as -1 and 0 still show up in the help block but as discussed those could be handled as edge cases later with relevant UXD

Thanks for the changes @abhinandan13jan 
/lgtm

nemesis09 commented:
/retest

nemesis09 commented:
/assign @andrewballantyne

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,196 @@
+
+  const handleRequestSizeChange = (size: RequestSize): void => {
+    const { value, unit } = size;
+    setRequestSizeValue(value);
```
Recommend you just handle the invalid values part of this PR as there is little reason to push it to a future issue.

Suggested change diff:
```
diff --git frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx
index 8f0b6f41e7..b359eb9eb7 100644
--- frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx
+++ frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx
@@ -42,6 +42,7 @@ const VolumeClaimTemplateForm: React.FC<VolumeClaimTemplateFormProps> = ({
   const [allowedAccessModes, setAllowedAccessModes] = React.useState(initialAccessModes);
   const [volumeMode, setVolumeMode] = React.useState(initialVolumeMode);
   const [accessMode, setAccessMode] = React.useState('ReadWriteOnce');
+  const [requestSizeError, setRequestSizeError] = React.useState<string>(null);
   const [requestSizeValue, setRequestSizeValue] = React.useState(initialSizeValue);
   const [requestSizeUnit, setRequestSizeUnit] = React.useState(initialSizeUnit);
   const [storageProvisioner, setStorageProvisioner] = React.useState('');
@@ -80,6 +81,14 @@ const VolumeClaimTemplateForm: React.FC<VolumeClaimTemplateFormProps> = ({

   const handleRequestSizeChange = (size: RequestSize): void => {
     const { value, unit } = size;
+
+    const numberValue = parseInt(value, 10);
+    if (Number.isNaN(numberValue) || numberValue <= 0) {
+      setRequestSizeError(t('pipelines-plugin~Size must be an integer greater than 0.'));
+      setRequestSizeValue('');
+      return;
+    }
+
     setRequestSizeValue(value);
     setRequestSizeUnit(unit);
   };
@@ -171,8 +180,13 @@ const VolumeClaimTemplateForm: React.FC<VolumeClaimTemplateFormProps> = ({
           describedBy="request-size-help"
           inputID="request-size-input"
           data-test-id="pvc-size-input"
-        />
-        <p className="help-block">{t('pipelines-plugin~Desired storage capacity')}</p>
+        >
+          {requestSizeError ? (
+            <p className="pf-c-form__helper-text pf-m-error">{requestSizeError}</p>
+          ) : (
+            <p className="help-block">{t('pipelines-plugin~Desired storage capacity')}</p>
+          )}
+        </RequestSizeInput>
         <label className="control-label" htmlFor="volume-mode">
           {t('pipelines-plugin~Volume Mode')}
         </label>
```

abhinandan13jan replied:
This will not disable the form unless the form field in set to an invalid value, I am doing that in line 98

andrewballantyne reviewed (changes requested):
Design has a gap between the dropdown and the alert for Empty Directory:
![Screen Shot 2021-03-25 at 9 38 07 AM](https://user-images.githubusercontent.com/8126518/112481909-df1ab600-8d4d-11eb-88b3-9aafd721334c.png)

Yours:
![Screen Shot 2021-03-25 at 9 39 45 AM](https://user-images.githubusercontent.com/8126518/112482084-10938180-8d4e-11eb-92e6-5c65b9a2374c.png)

There is also a `.` at the end of the Alert sentence.

---

The dropdown items should match the design. You have an odd gap in your `VolumeClaim Template`:
![Screen Shot 2021-03-25 at 9 39 36 AM](https://user-images.githubusercontent.com/8126518/112482166-22752480-8d4e-11eb-8b09-e24ec44e5485.png)

Design:
![Screen Shot 2021-03-25 at 9 39 13 AM](https://user-images.githubusercontent.com/8126518/112482185-27d26f00-8d4e-11eb-9025-84147da8451f.png)

Could you also convert the PVC to the fully qualified `PersistentVolumeClaim`? Also if it's easy to do, please rearrange the order to match designs, I think that's cleaner imo.

---

It's very tight on your form, can you space out the fields a little more? See designs:
![Screen Shot 2021-03-25 at 9 43 20 AM](https://user-images.githubusercontent.com/8126518/112482562-96173180-8d4e-11eb-9b4e-74c6838caffd.png)

Yours for comparison:
![Screen Shot 2021-03-25 at 9 43 10 AM](https://user-images.githubusercontent.com/8126518/112482596-9c0d1280-8d4e-11eb-8af1-2ac1aa047bec.png)

abhinandan13jan commented:
@andrewballantyne I've changed to make the form aligned to designs and updated the latest screenshots. please revisit
![Screenshot from 2021-03-30 16-16-05](https://user-images.githubusercontent.com/24852534/112978074-7d56b380-9174-11eb-849b-a5e3e4665796.png)
![Screenshot from 2021-03-30 16-16-13](https://user-images.githubusercontent.com/24852534/112978077-7e87e080-9174-11eb-82aa-6cb071ba6804.png)

andrewballantyne commented:
@abhinandan13jan I don't see a new push? I think you forgot to push the changes...

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelines/modals/common/VolumeClaimTemplateForm.tsx:
```diff
@@ -0,0 +1,213 @@
+        textExpanded={t('pipelines-plugin~Hide VolumeClaimTemplate options')}
+        textCollapsed={t('pipelines-plugin~Show VolumeClaimTemplate options')}
+      >
+        <p>
```
You shouldn't use `p` tags as a means to just force content separation. It's not the intended use of the paragraph tag. If you elements were inline, then it could possibly be used, but still not in the case of a form.

Recommend you just change them to be `div` and add a className of `opp-VolumeClaimTemplateForm--element` or something along that lines. Create a bottom margin/padding of 8px or 4px or whatever is needed to make it similar to the design.

andrewballantyne reviewed (changes requested):
@abhinandan13jan You didn't address any of these...

Design has a gap between the dropdown and the alert for Empty Directory:
![Screen Shot 2021-03-25 at 9 38 07 AM](https://user-images.githubusercontent.com/8126518/112481909-df1ab600-8d4d-11eb-88b3-9aafd721334c.png)

Yours:
![Screen Shot 2021-03-25 at 9 39 45 AM](https://user-images.githubusercontent.com/8126518/112482084-10938180-8d4e-11eb-92e6-5c65b9a2374c.png)

There is also a `.` at the end of the Alert sentence.

---

The dropdown items should match the design. You have an odd gap in your `VolumeClaim Template`:
![Screen Shot 2021-03-25 at 9 39 36 AM](https://user-images.githubusercontent.com/8126518/112482166-22752480-8d4e-11eb-8b09-e24ec44e5485.png)

Design:
![Screen Shot 2021-03-25 at 9 39 13 AM](https://user-images.githubusercontent.com/8126518/112482185-27d26f00-8d4e-11eb-9025-84147da8451f.png)

Could you also convert the PVC to the fully qualified `PersistentVolumeClaim`? Also if it's easy to do, please rearrange the order to match designs, I think that's cleaner imo.

abhinandan13jan commented:
> @abhinandan13jan You didn't address any of these...
> 
> Design has a gap between the dropdown and the alert for Empty Directory:
> ![Screen Shot 2021-03-25 at 9 38 07 AM](https://user-images.githubusercontent.com/8126518/112481909-df1ab600-8d4d-11eb-88b3-9aafd721334c.png)
> 
> Yours:
> ![Screen Shot 2021-03-25 at 9 39 45 AM](https://user-images.githubusercontent.com/8126518/112482084-10938180-8d4e-11eb-92e6-5c65b9a2374c.png)
> 
> There is also a `.` at the end of the Alert sentence.
> 
> The dropdown items should match the design. You have an odd gap in your `VolumeClaim Template`:
> ![Screen Shot 2021-03-25 at 9 39 36 AM](https://user-images.githubusercontent.com/8126518/112482166-22752480-8d4e-11eb-8b09-e24ec44e5485.png)
> 
> Design:
> ![Screen Shot 2021-03-25 at 9 39 13 AM](https://user-images.githubusercontent.com/8126518/112482185-27d26f00-8d4e-11eb-9025-84147da8451f.png)
> 
> Could you also convert the PVC to the fully qualified `PersistentVolumeClaim`? Also if it's easy to do, please rearrange the order to match designs, I think that's cleaner imo.

@andrewballantyne I've updated. need re-review

andrewballantyne commented:
We have a Enum problem that finally caught up with us. Please accept my PR on your fork/branch: https://github.com/abhinandan13jan/console/pull/5

It'll fix this issue:
![Screen Shot 2021-03-31 at 11 17 14 AM](https://user-images.githubusercontent.com/8126518/113184743-92d6e680-9223-11eb-92dd-9cbf5a23bf0c.png)

abhinandan13jan commented:
> We have a Enum problem that finally caught up with us. Please accept my PR on your fork/branch: [abhinandan13jan#5](https://github.com/abhinandan13jan/console/pull/5)
> 
> It'll fix this issue:

Merged your PR @andrewballantyne

christianvogt commented on frontend/packages/pipelines-plugin/src/components/pipelines/const.ts:
```diff
@@ -28,13 +28,21 @@ export const pipelineResourceTypeSelections = {
+  VolumeClaimTemplate = 'volumeClaimTemplate',
 }
 
+export const volumeTypeOptions: { [type in VolumeTypes]: string } = {
+  [VolumeTypes.EmptyDirectory]: 'Empty Directory',
+  [VolumeTypes.ConfigMap]: 'Config Map',
+  [VolumeTypes.Secret]: 'Secret',
+  [VolumeTypes.PVC]: 'PersistentVolumeClaim',
+  [VolumeTypes.VolumeClaimTemplate]: 'VolumeClaimTemplate',
+};
```
What about localization?

andrewballantyne replied:
Whoops... guess that should be done too!

abhinandan13jan replied:
updated

rottencandy reviewed (approved):
/lgtm

nemesis09 commented:
/lgtm
